More integration tests for CLI (including minor fixes in CLI)

piotr-iohk opened the pull request:
# Issue Number

#96
<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->


# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [ ] I have added more tests cases for CLI (including tests for testing wallet creation)
- [ ] Did minor fixes in CLI: 
 - making display of help with consistent exit code in `cardano-wallet-launcher` and `cardano-wallet` for both parameters (`--help` and `-h`)
 - making `mnemonic generate` to always show `Invalid mnemonic size. Expected one of: 9,12,15,18,21,24` for invalid `--size` (it used to show rather unpleasent `Int is an integer number between -9223372036854775808 and 9223372036854775807.` when the provided `--size` was not Int)
- [ ] did slight refactoring (moving data used in test to separate module `TestData.hs` and extracting methods for interacting with CLI to `DSL.hs`)


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

denisshevchenko commented on src/Cardano/Wallet/WalletLayer/Kernel/Conv.hs:
```diff
@@ -180,6 +181,23 @@ toWallet db hdRoot = V1.Wallet {
+      eoswalId             = V1.WalletId walletId
+    , eoswalName           = hdRoot ^. HD.hdRootName
+                                     . to HD.getWalletName
+    , eoswalAddressPoolGap = error "TODO, we have to take it from one of accounts, will be done later, see issue #36"
```
We store address pool gap in `HdAccount`s, not in `HdRoot`, so we must take any of corresponding accounts and take gap from it. It will be done later.

KtorZ replied:
What about doing it as part of this PR :) ?

denisshevchenko replied:
@KtorZ Done (please see last commit).

denisshevchenko commented on src/Cardano/Wallet/WalletLayer.hs:
```diff
@@ -168,19 +167,6 @@ instance Buildable DeleteWalletError where
-instance Buildable DeleteEosWalletError where
-    build (DeleteEosWalletError kernelError) =
-        bprint ("DeleteEosWalletError " % build) kernelError
-
```
I thought that we can use the same `DeleteWalletError`, because we actually delete the same `HdRoot` as for EO-wallets.

KtorZ commented on src/Cardano/Wallet/WalletLayer/Kernel/Conv.hs:
```diff
@@ -159,26 +160,51 @@ toAccount snapshot account = V1.Account {
+    createdAt  = hdRoot ^. HD.hdRootCreatedAt . fromDb
+    level      = hdRoot ^. HD.hdRootAssurance
+    walletName = hdRoot ^. HD.hdRootName . to HD.getWalletName
+    walletId   = sformat build rootId
```
I'd rather not have that function. Having some similar where clauses in both `toEosWallet` and `toWallet` is fine...

denisshevchenko replied:
`hlint` is angry in this case (Duplication). :-)

KtorZ replied:
Well, that wouldn't be the first case `hlint` suggests something wrong :upside_down_face:

KtorZ replied:
Unless you truly think and argue that this makes the code looks better ...

KtorZ replied:
So we've reduce duplication to avoid having to modify code in two different places. But now, we'd have to modify it in 3 different places. Not sure what we've improved here to be honest. 

In order to really remove duplication, we could have provided a different `toWallet` function that may just construct the first part common part of an `EosWallet` or `Wallet` and leave the rest to the caller... But to be honest, this complexifies the whole for close to no value. 

So yes, I am arguably saying that `hlint` can go to hell here :)

KtorZ commented on src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs:
```diff
@@ -242,6 +248,36 @@ getWallet wallet wId db = runExceptT $ do
+    -- Since this function is using for EOS-wallets only,
+    -- we definitely know that this is an account with EO-branch.
+    HD.HdAccountBaseEO _ _ gap = firstAccount ^. HD.hdAccountBase
+    (firstAccount:_) = IxSet.toList accountSet
```
Hmm. This feels wrong and may eventually bite us. 
We should:

- Forbid empty E.O. wallets creation (not in this PR, but I'll add a bullet point to #236)
- Make sure that all accounts have a same gap within a wallet
- Return an error here if there's no account

denisshevchenko replied:
> Make sure that all accounts have a same gap within a wallet

Of course they have the same gap, because they was _created_ from the same gap (during EOS-wallet creation).

KtorZ replied:
> Of course they have the same gap,

They **ought to** have the same gap. But there's nothing that can be checked at compile-time that guarantees this invariant. Hence the need for checking it and, throwing an exception if it doesn't hold. 

>  In this case we just return default value of gap.

No, that's the point. If there's no accounts, that's an error! This shouldn't happen and we should enforce that via the API. If it ever happens, that's a bug / fault in our implementation

KtorZ commented on src/Cardano/Wallet/WalletLayer/Kernel/Conv.hs:
```diff
@@ -180,6 +186,23 @@ toWallet db hdRoot = V1.Wallet {
+toEosWallet gap db hdRoot = V1.EosWallet {
+      eoswalId             = V1.WalletId walletId
+    , eoswalName           = hdRoot ^. HD.hdRootName
+                                     . to HD.getWalletName
```
Please, no more variable-length indent :pray:

KtorZ commented on src/Cardano/Wallet/WalletLayer/Kernel/Conv.hs:
```diff
@@ -62,6 +64,10 @@ import qualified Cardano.Wallet.Kernel.Read as Kernel
 -- import           Cardano.Wallet.WalletLayer (InvalidRedemptionCode (..))
 
+-- Functions 'toWallet' and 'toEosWallet' contain duplications in 'where'
+-- sections, but this is not a problem.
```
This doesn't really explain _why_ it's not a problem :/ ..

denisshevchenko commented:
@KtorZ I've added checks for `HdAccount`s in EOS-wallets, with corresponding errors.

denisshevchenko commented on src/Cardano/Wallet/API/V1/ReifyWalletError.hs:
```diff
@@ -156,6 +157,19 @@ getWalletError e = case e of
+    ex@(GetEosWalletErrorWrongAccounts _txt) ->
+        V1.UnknownError (sformat build ex)
+    ex@(GetEosWalletErrorGapsDiffer _txt) ->
+        V1.UnknownError (sformat build ex)
```
Thanks, fixed.

KtorZ commented on src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs:
```diff
@@ -242,6 +252,38 @@ getWallet wallet wId db = runExceptT $ do
+    :: HD.HdRootId
+    -> Kernel.DB
+    -> Either GetEosWalletError AddressPoolGap
+addressPoolGapByRootId rootId db = do
```
So ideally, this would return a specific error as well. Truth is that the branches below of `GetEosWalletError` are only met within this function. And, other branches can never be met by this function.
So, saying that this function can return a `WalletIdDecodingFailed` is a bit of a lie :sweat_smile: ...

denisshevchenko replied:
Done.

KtorZ commented on src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs:
```diff
@@ -259,6 +301,22 @@ getWallets wallet db =
+    let (errors, eosWallets) = partitionEithers result
+    return $ if not . null $ errors
+        then let (err:_) = errors in Left err
+        else Right $ IxSet.fromList eosWallets
```
A case reads better for this kind of manipulation :+1:

denisshevchenko replied:
You're right! Fixed.

KtorZ commented on src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs:
```diff
@@ -259,6 +301,22 @@ getWallets wallet db =
+    result <- forM allRoots $ \root -> runExceptT $
+        fmap (toEosWallet db root) $ exceptT $
+            addressPoolGapByRootId (root ^. HD.hdRootId) db
+    let (errors, eosWallets) = partitionEithers result
```
We aren't quite leveraging the applicative aspect of `Either` and `ExceptT` here. Especially if we just intend to stop at the first error, we could instead `traverse` the list of hd roots and end up with a `Either Error [EosWallet]`

denisshevchenko replied:
Done.

denisshevchenko commented:
@KtorZ All checks have passed (`hlint` is happy again).

KtorZ commented on src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs:
```diff
@@ -259,6 +302,21 @@ getWallets wallet db =
+    -> Kernel.DB
+    -> m (Either GetEosWalletError (IxSet V1.EosWallet))
+getEosWallets _wallet db = do
+    let result = traverse (\root -> either Left (Right . toEosWallet db root) $
```
Note that:

```hs
\root -> either Left (Right . toEosWallet db root)
````

can be written more idiomatically as:

```
fmap (toEosWallet db)
```

Commit:
Make cardano-wallet and cardano-wallet-launcher help consistent

Commit:
tests for mnemonic generate --size and making error message to be awlays meaningful in case of invalid --size

Commit:
- introduced ability to test creation of the wallet via CLI
 - moving test data to separate module TestData.hs
 - extracting methods for interacting with CLI to DSL
 - few more test cases

piotr-iohk commented on lib/http-bridge/test/integration/Test/Integration/Scenario/CLISpec.hs:
```diff
@@ -31,72 +31,167 @@ import Test.Integration.Framework.DSL
+            if (title == "40 chars hex") then
+                c `shouldBe` ExitSuccess
+            else
+                c `shouldBe` ExitFailure 1
```
This is slightly inconsistent I think. I suppose the response code should be always `1` (currently for valid hex, but not existing wallet it is `0`, for others it is `1`)

KtorZ replied:
Hmmm... That's actually right.

Commit:
listAddressesViaCLI

KtorZ commented on exe/wallet/Main.hs:
```diff
@@ -343,15 +347,15 @@ execServer (Port port) (Port bridgePort) = do
+        "15" -> mnemonicToText @15 . entropyToMnemonic <$> genEntropy
+        "18" -> mnemonicToText @18 . entropyToMnemonic <$> genEntropy
+        "21" -> mnemonicToText @21 . entropyToMnemonic <$> genEntropy
+        "24" -> mnemonicToText @24 . entropyToMnemonic <$> genEntropy
```
Sweet :o

KtorZ commented on lib/http-bridge/test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -438,3 +462,54 @@ wantedErrorButSuccess
+                hPutStr i (passphrase ++ "\n")
+                hPutStr i (passphrase ++ "\n")
+                hFlush i
+                hClose i
```
My man :heart:

KtorZ commented on lib/http-bridge/test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -438,3 +462,54 @@ wantedErrorButSuccess
+            { std_in = CreatePipe, std_out = CreatePipe, std_err = CreatePipe }
+    withCreateProcess process $ \stdIn _ _ h -> do
+        case stdIn of
+            Nothing -> return ()
```
I wonder if this should "throw". Since we use `CreatePipe` this _must_ be `Just`. If not, the test might hang forever.
